File: leetcode/dp/2297_Jump Game_VIII/codes/top_down.py
from functools import cache
from typing import List

class Solution:
    def minCost(self, nums: List[int], costs: List[int]) -> int:
        n = len(nums)
        nse, nge = [n] * n, [n] * n
        max_st, min_st = [], []

        for i in range(n):
            while max_st and nums[max_st[-1]] <= nums[i]:
                nge[max_st.pop()] = i
            while min_st and nums[min_st[-1]] > nums[i]:
                nse[min_st.pop()] = i
            max_st.append(i)
            min_st.append(i)

        @cache
        def dp(i):
            if i == n:
                return float('inf')
            if i == n - 1:
                return 0 if i == 0 else costs[i]
            cost = 0 if i == 0 else costs[i]
            return cost + min(dp(nse[i]), dp(nge[i]))
        
        return dp(0)

File: leetcode/dp/2297_Jump Game_VIII/codes/top_down.cpp
 #include <vector>
#include <stack>
#include <climits>
#include <algorithm>
using namespace std;

class Solution {
public:
    long long dp(int i, vector<int>& nums, vector<int>& costs, vector<int>& nse, vector<int>& nge, vector<long long>& memo, int n) {
        if (i == n) return LLONG_MAX;
        if (memo[i] != -1) return memo[i];
        if (i == n - 1) return memo[i] = (i == 0 ? 0 : costs[i]);
        long long cost = (i == 0 ? 0 : costs[i]);
        long long left = dp(nse[i], nums, costs, nse, nge, memo, n);
        long long right = dp(nge[i], nums, costs, nse, nge, memo, n);
        return memo[i] = cost + min(left, right);
    }

    long long minCost(vector<int>& nums, vector<int>& costs) {
        int n = nums.size();
        vector<int> nse(n, n), nge(n, n);
        vector<long long> memo(n, -1);
        stack<int> max_st, min_st;

        for (int i = 0; i < n; ++i) {
            while (!max_st.empty() && nums[max_st.top()] <= nums[i]) {
                nge[max_st.top()] = i;
                max_st.pop();
            }
            while (!min_st.empty() && nums[min_st.top()] > nums[i]) {
                nse[min_st.top()] = i;
                min_st.pop();
            }
            max_st.push(i);
            min_st.push(i);
        }

        return dp(0, nums, costs, nse, nge, memo, n);
    }
};
